Use builtin OpenAI plugin. Adds tool support.

kanaka opened the pull request:
The GitHub Copilot completions (`/chat/completions`) API is based on the OpenAI API regardless of whether the backend model is an OpenAI one or not. This means we can greatly simplify the GitHub Copilot plugin by using the builtin OpenAI plugin (`llm.default_plugins.openai_models`) for everything except for the Copilot device-based authentication.

In addition to the significant simplification and code reduction, one key benefit of this approach is that we automatically get additional functionality that is implemented by the openai_models plugin such as tool calling, async, schema/structured output, etc. We should also get any future OpenAI plugin functionality automatically (or with minimal changes).

Tests have been updated to remove most of the completion tests since this should now be covered by tests in OpenAI plugin itself. Minor updates were needed to the authentication and the CLI tests to get them to work simplified code base.

Finally, the fetch_models_data top-level function (wraps _fetch_models_data) has been memoized so that the model data is cached after the first call.

NOTES:
* To support tool calls in non OpenAI models, this change is needed in llm itself: https://github.com/simonw/llm/pull/1170 However, this change isn't dependent on that one (but once it's merge tool calls will work in anthropic models too).
* I haven't bumped the version as part of this PR but that should happen before this is released to pypi of course.
* I removed the use of "github_copilot" as a default model. Trying to keep it was increasing the complexity of the result quite a bit. However, I don't feel strongly about that if somebody else wants to add it back.

kanaka commented:
@jmdaly can you kick off the workflow again to see if my changes fixed the 3 test failures?

jmdaly commented:
> @jmdaly can you kick off the workflow again to see if my changes fixed the 3 test failures?

Done!

kanaka commented:
@jmdaly Ah, apparently my existing auth config was allowing those three tests to pass. I'm pretty confident I've fixed those test issues now (an additional monkey patch in the auth path). I've also added a Dockerfile and test script that makes it easier to test in an uncontaminated environment. The script `run-tests.sh` does the same linting and test steps in the testing workflow (could be used to replace a couple of those workflow steps if you want).

Could you kick off the workflow again?

kanaka commented:
@jmdaly You might be interested in this post where I leverage llm and this branch of llm-github-copilot to show how to build your own coding agent: https://kanaka.github.io/blog/llm-agent-in-five-steps/

kanaka commented:
@jmdaly I rebased to address the merge conflicts from the other PR merge. I also just pushed a tiny cleanup to the class inheritance that I discovered.

The resulting diff is a little bit hard to read. Here's the tl;dr version:
* `GitHubCopilotAuthenticator` class is unchanged.
* Replace `GithubCopilot` class with sync and async classes that wrap the builtin openai classes and that just override auth key, copilot api_base, and capabilities info
* Simplify how model data is fetched (and add caching/memoizing to it)
* Drop the default model handling conditions since it wasn't immediately clear how this should integrate with openai builtin classes.

jmdaly commented:
Also, removing almost 800 lines is awesome :)

cavanaug commented:
Let me look into this in general and test it a bit, but I really do like the concept of reusing an existing library and automatically gaining benefits.   Though for tool calling we will want to stay in alignment with whatever Simon does with llm in general, we dont want to expose any additional potential security implications with just this plugin.

kanaka commented:
@cavanaug Hopefully this should make it a lot easier to stay in sync with functionality in llm (including tools), because with this change, the plugin now only does auth and model listing (https://github.com/jmdaly/llm-github-copilot/blob/d9e1202454b6b4fe1164611f4c4196327f25d743/llm_github_copilot.py#L476-L509). All the other API calls and functionality are handled by the builtin openai plugin (that Simon maintains in llm). Any tool policy or security issues would be handled in the openai module and be transparent to the llm-github-copilot plugin. Thinking pretty far out of the box, there is a weird case where a tool is trying to query auth settings, but I would expect that to just break things rather than expose a vulnerability because the plugin doesn't do anything except set the api url and auth key during __init__.

cavanaug commented:
I spent some time looking at this and testing it locally.   LGTM.    Great simplification!!

jmdaly commented:
> Yay! @jmdaly when you have a chance can you update the version and push a release? Thanks for your work on this plugin!

Yes definitely! I'll try to get it done this weekend :) Thanks again for your great contribution!
